nilearn/regions/hierarchical_kmeans_clustering.py

```diff
@@ -0,0 +1,276 @@
+        self.n_init = n_init
+        self.max_no_improvement = max_no_improvement
+        self.verbose = verbose
+        self.random_state = random_state
+        self.scaling = scaling
+
+    def fit(self, X, y=None):
+        """Compute clustering of the data.
+
+        Parameters
+        ----------
+        X: ndarray, shape = [n_features, n_samples]
+            Training data.
+        y: Ignored
+
+        Returns
+        -------
+        self
+        """
+
+        X = check_array(X, ensure_min_features=2, ensure_min_samples=2,
+                        estimator=self)
+        n_features = X.shape[1]
+
+        if self.n_clusters <= 0:
+            raise ValueError("n_clusters should be an integer greater than 0."
+                             " %s was provided." % str(self.n_clusters))
+
+        if self.n_clusters > n_features:
+            self.n_clusters = n_features
+            warnings.warn("n_clusters should be at most the number of "
+                          "features. Taking n_clusters = %s instead."
+                          % str(n_features))
+        self.labels_ = hierarchical_k_means(X, self.n_clusters, self.init,
+                                            self.batch_size, self.n_init,
+                                            self.max_no_improvement,
+                                            self.verbose, self.random_state)
+        sizes = np.bincount(self.labels_)
+        sizes = sizes[sizes > 0]
+        self.sizes_ = sizes
+        self.n_clusters = len(np.unique(self.labels_))
+        return self
+
+    def transform(self, X, y=None):
+        """Apply clustering, reduce the dimensionality of the data.
+
+        Parameters
+        ----------
+        X: ndarray, shape = [n_features, n_samples]
+            Data to transform with the fitted clustering.
+
+        Returns
+        -------
+        X_red: ndarray, shape = [n_clusters, n_samples]
+            Data reduced with agglomerated signal for each cluster
+        """
+
+        check_is_fitted(self, "labels_")
+        unique_labels = np.unique(self.labels_)
+
+        mean_cluster = []
+        for label in unique_labels:
+            mean_cluster.append(np.mean(X[self.labels_ == label], axis=0))
+
+        X_red = np.array(mean_cluster)
+
+        if self.scaling:
+            X_red = X_red * np.sqrt(self.sizes_[:, np.newaxis])
+
+        return X_red
+
+    def inverse_transform(self, X_red):
+        """Send the reduced 2D data matrix back to the original feature
+        space (voxels).
+
+        Parameters
+        ----------
+        X_red: ndarray , shape = [n_clusters, n_samples]
+            Data reduced with agglomerated signal for each cluster
+
+        Returns
+        -------
+        X_inv: ndarray, shape = [n_features, n_samples]
+            Data reduced expanded to the original feature space
+        """
+
+        check_is_fitted(self, "labels_")
+
+        _, inverse = np.unique(self.labels_, return_inverse=True)
```

Comment:
iiuc the labels corresponding to reduced dimensions will be 0, ... n_clusters -1 so this might not be necessary? and if it is maybe we can store it after fit